### PR TITLE
[SPARK-41412][CONNECT] Implement `Column.cast`

### PR DESCRIPTION
**Pull request by amaliujia**
### What changes were proposed in this pull request?

1. Add `types.py` and move to/from type proto to this file.
2. Improve Cast proto to support both DataType and DataType as a string.
3.  support cast to the following types in Python:
    ByteType,
    ShortType,
    IntegerType,
    FloatType,
    DayTimeIntervalType,
    MapType,
    StringType,
    DoubleType,
    LongType,
    DecimalType,
    BinaryType,
    BooleanType

### Why are the changes needed?

API coverage

### Does this PR introduce _any_ user-facing change?

NO

### How was this patch tested?

UT

### CODE REVIEW

**Comment by amaliujia** [comment]
cc @zhengruifeng

File: connector/connect/common/src/main/protobuf/spark/connect/expressions.proto [hunk 1]
```diff
@@ -43,7 +43,11 @@ message Expression {
     Expression expr = 1;
 
     // (Required) the data type that the expr to be casted to.
-    DataType cast_to_type = 2;
+    oneof cast_to_type {
```
Comment (grundprinzip) [inline connector/connect/common/src/main/protobuf/spark/connect/expressions.proto:46]: FYI: this is a breaking change in the proto message. Ideally, we would use

```
reserved 2;
oneof cast_to_type {
  DataType type = 3;
  string type_str = 4;
}
```

Comment (amaliujia) [reply connector/connect/common/src/main/protobuf/spark/connect/expressions.proto:46]: Ah thanks. I didn't know this way to evolve the proto.

Given Spark Connect is still alpha component though for now we don't need to be enforced to maintain the backwards compatibility

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 2]
```diff
@@ -518,9 +518,16 @@ class SparkConnectPlanner(session: SparkSession) {
...
-      DataTypeProtoConverter.toCatalystType(cast.getCastToType))
+    cast.getCastToTypeCase match {
+      case proto.Expression.Cast.CastToTypeCase.TYPE =>
+        Cast(
+          transformExpression(cast.getExpr),
+          DataTypeProtoConverter.toCatalystType(cast.getType))
+      case _ =>
+        Cast(
+          transformExpression(cast.getExpr),
+          session.sessionState.sqlParser.parseDataType(cast.getTypeStr))
```
[the remark this answers is not in the document]
Comment (amaliujia) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:529]: I am not following this? Can you give some example proto/sample code?

File: connector/connect/common/src/main/protobuf/spark/connect/expressions.proto (hunk 1 above)
Comment (zhengruifeng) [reply connector/connect/common/src/main/protobuf/spark/connect/expressions.proto:46]: I think it's fine since this message has not been actually used?

Comment (grundprinzip) [reply connector/connect/common/src/main/protobuf/spark/connect/expressions.proto:46]: We're ok now. We can use Buf to check for breaking changes.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (grundprinzip) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:529]: When I looked at cast as a unicorn in the context of expressions, it's one of the few where not all argument types resolve to expressions. I was wondering if we could simplify the approach to make the type argument of cast an expression that can resolve as a string then we can do the matching of the expression in the analyzer.

This is very similar to why you added the oneof to the proto message.

**Review by grundprinzip — APPROVED** [review]

**Comment by grundprinzip** [comment]
My discussion was not meant to block the PR, this was more a more general observation. 

LGTM, thanks!

cc @HyukjinKwon @zhengruifeng

**Comment by zhengruifeng** [comment]
merged into master

**Pull request closed without merging**